### PR TITLE
Store to local for spark

### LINKED ISSUE #327: Alternate method to communicate variables between livy and local (not just using %%sql -o)
It would be great if we also could communicate in some way the contents of a dataframe/pandas in code, not just using the `%sql -o` way.

### PR DESCRIPTION
**Pull request by orestis-ms**
Closes #327 

Possible directions from here:
1. Allow the cell to be executed without any code (make line-magic instead of cell-magic)
2. Refactor SQLQuery, Command and the newly created SparkStoreCommand since they are all 'commands' and all inherit (directly or indirectly) from ObjectWithGuid.

### CODE REVIEW

File: sparkmagic/sparkmagic/magics/sparkmagicsbase.py [hunk 1]
```diff
@@ -34,6 +37,21 @@ def __init__(self, shell, data=None, spark_events=None):
...
 
+    def execute_spark(self, cell, output_var, samplemethod, maxrows, samplefraction, session_name):
+        (success, out) = self.spark_controller.run_command(Command(cell), session_name)
+        if not success:
+            self.ipython_display.send_error(out)
+        else:
+            self.ipython_display.write(out)
+            if output_var is not None:
+                spark_store_command = self._spark_store_command(cell, output_var, samplemethod, maxrows, samplefraction)
+                df = self.spark_controller.run_command(spark_store_command, session_name)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/magics/sparkmagicsbase.py]: What if the spark_store_command is not successful: e.g. variable doesn't exist?

Comment (aggFTW) [inline sparkmagic/sparkmagic/magics/sparkmagicsbase.py]: As it is, we would simply be catching the exception and displaying it, but I wonder if there's a possibility for a better error message... Ideas?

File: sparkmagic/sparkmagic/tests/test_kernel_magics.py [hunk 2]
```diff
@@ -502,10 +502,50 @@ def test_spark_expected_exception():
...
+    magic.spark(line, cell)
+    assert spark_controller.run_command.call_count == 2
+    spark_controller.run_command.assert_any_call(Command(cell), None)
+    ipython_display.send_error.assert_called_with(constants.INTERNAL_ERROR_MSG
+                                                       .format(side_effect[1]))
+
+
+@with_setup(_setup, _teardown)
+def test_spark_expected_exception_in_storing():
+    line = "-o var_name"
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_kernel_magics.py]: This test's name is the same as the one above. Please change the one above to unexpected.

File: sparkmagic/sparkmagic/tests/test_kernel_magics.py [hunk 3]
```diff
@@ -502,10 +502,50 @@ def test_spark_expected_exception():
...
+
+@with_setup(_setup, _teardown)
+def test_spark_sample_options():
+    line = "-o var_name -m sample -n 142 -r 0.3"
+    cell = ""
+    magic.execute_spark = MagicMock()
+
+    ret = magic.spark(line, cell)
+
+    spark_controller.add_session.assert_called_once_with(magic.session_name, magic.endpoint, False,
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_kernel_magics.py]: Why are we checking that `add_session` executes?

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py [hunk 4]
```diff
@@ -269,6 +269,99 @@ def test_run_cell_command_exception():
...
+    meth = "-m"
+    method_name = "sample"
+    output = "-o"
+    output_var = "var_name"
+    line = " ".join([command, name, context, context_name, meth, method_name, output, output_var])
+    cell = "cell code"
+
+    result = magic.spark(line, cell)
+
+    run_cell_method.assert_any_call(Command(cell), name)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: Assert that the result of Command (`""` in this unit test) is printed to the user.

File: sparkmagic/sparkmagic/tests/test_sparkmagicbase.py [hunk 5]
```diff
@@ -131,3 +131,46 @@ def test_print_empty_endpoint_info():
     magic._print_endpoint_info([], current_session_id)
     magic.ipython_display.html.assert_called_once_with(u'No active sessions.')
 
+
+def test_spark_execution_without_output_var():
+    magic = SparkMagicBase(None)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkmagicbase.py]: It seems like the setup can be done in a common `setup` method.

File: sparkmagic/sparkmagic/tests/test_sparkmagicbase.py [hunk 6]
```diff
@@ -131,3 +131,46 @@ def test_print_empty_endpoint_info():
...
+    shell = MagicMock()
+    shell.user_ns = {}
+    magic.shell = shell
+
+    session = MagicMock()
+    output_var = "var_name"
+    code = ""
+    df = 'df'
+
+    magic.spark_controller.run_command.side_effect = [(True,'out'), df]
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkmagicbase.py]: We're missing a case here where the second command is an exception.

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py [hunk 7]
```diff
@@ -269,6 +269,99 @@ def test_run_cell_command_exception():
...
+    ipython_display.send_error.assert_called_once_with(EXPECTED_ERROR_MSG
+                                                       .format(run_cell_method.side_effect))
+
+@with_setup(_setup, _teardown)
+def test_run_spark_command_exception_while_storing():
+    run_cell_method = MagicMock()
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: This unit test should be in SparkMagicBase tests

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py [hunk 8]
```diff
@@ -269,6 +269,99 @@ def test_run_cell_command_exception():
...
+
+    result = magic.spark(line, cell)
+
+    run_cell_method.assert_any_call(Command(cell), name)
+    run_cell_method.assert_any_call(SparkStoreCommand(cell, output_var, samplemethod=method_name), name)
+
+
+@with_setup(_setup, _teardown)
+def test_run_spark_command_exception():
+    run_cell_method = MagicMock()
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: This unit test should be in SparkMagicBase tests

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py [hunk 9]
```diff
@@ -269,6 +269,99 @@ def test_run_cell_command_exception():
...
+    meth = "-m"
+    method_name = "sample"
+    output = "-o"
+    output_var = "var_name"
+    line = " ".join([command, name, context, context_name, meth, method_name, output, output_var])
+    cell = "cell code"
+
+    result = magic.spark(line, cell)
+
+    run_cell_method.assert_any_call(Command(cell), name)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: We should just assert that we are calling the base method.

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py [hunk 10]
```diff
@@ -269,6 +269,99 @@ def test_run_cell_command_exception():
...
+    context = "-c"
+    context_name = "spark"
+    meth = "-m"
+    method_name = "sample"
+    line = " ".join([command, name, context, context_name, meth, method_name])
+    cell = "cell code"
+
+    result = magic.spark(line, cell)
+
+    run_cell_method.assert_called_once_with(Command(cell), name)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: We should just assert that we are calling the base method.

File: sparkmagic/sparkmagic/tests/test_sparkmagicbase.py [hunk 11]
```diff
@@ -131,3 +131,46 @@ def test_print_empty_endpoint_info():
...
+    magic = SparkMagicBase(None)
+    magic.spark_controller = MagicMock()
+    magic.ipython_display = MagicMock()
+
+    session = MagicMock()
+    output_var = None
+    
+    magic.spark_controller.run_command.return_value = (True,'out')
+    magic.execute_spark("", output_var, None, None, None, session)
+    magic.ipython_display.write.assert_called_once_with('out')
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkmagicbase.py]: Assert that run_cell_method is not called with SparkStoreCommand

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+def test_to_command_pyspark():
+    variable_name = "var_name"
+    sparkcommand = SparkStoreCommand("Code", variable_name)
+    sparkcommand._pyspark_command = MagicMock(return_value=MagicMock())
+    sparkcommand.to_command("pyspark", variable_name)
+    sparkcommand._pyspark_command.assert_called_with(variable_name)
+
+
+@with_setup(_setup, _teardown)
+def test_to_command_pyspark3():
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: test_to_command for scala and R?

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+
+
+@with_setup(_setup, _teardown)
+def test_sparkstorecommand_loads_defaults():
+    defaults = {
+        conf.default_samplemethod.__name__: "sample",
+        conf.default_maxrows.__name__: 419,
+        conf.default_samplefraction.__name__: 0.99,
+    }
+    conf.override_all(defaults)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: You need to undo the override on `_teardown` above

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+    variable_name = "var_name"
+    code = "hugedataset.collect()"
+    sparkcommand = SparkStoreCommand(code, variable_name)
+    assert_equals(sparkcommand.code, code)
+    assert_equals(sparkcommand.samplemethod, defaults[conf.default_samplemethod.__name__])
+    assert_equals(sparkcommand.maxrows, defaults[conf.default_maxrows.__name__])
+    assert_equals(sparkcommand.samplefraction, defaults[conf.default_samplefraction.__name__])
+
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Nit: Add extra line

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+    code = "hugedataset.collect()"
+    samplemethod = "foo"
+    _ = SparkStoreCommand(code, variable_name, samplemethod)
+
+
+@with_setup(_setup, _teardown)
+def test_pyspark_livy_sampling_options():
+    variable_name = "abc"
+    code = "hugedataset.collect()"
+    sparkcommand = SparkStoreCommand(code, variable_name, samplemethod='take', maxrows=120)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: We are missing a test for when sample method is `None` and it should do a `collect` for all 4 languages

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+                                  LONG_RANDOM_VARIABLE_NAME, conf.pyspark_python_encoding())))
+
+    sparkcommand = SparkStoreCommand(code, variable_name, samplemethod='sample', samplefraction=0.33, maxrows=3234)
+    assert_equals(sparkcommand._pyspark_command(variable_name),
+                  Command(u'for {} in {}.toJSON().sample(False, 0.33).take(3234): '
+                          u'print({}.encode("{}"))'\
+                          .format(LONG_RANDOM_VARIABLE_NAME, variable_name,
+                                  LONG_RANDOM_VARIABLE_NAME, conf.pyspark_python_encoding())))
+
+    sparkcommand = SparkStoreCommand(code, variable_name, samplemethod='sample', samplefraction=0.33, maxrows=3234)
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: duplicated

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+import sparkmagic.utils.configuration as conf
+import sparkmagic.utils.constants as constants
+from sparkmagic.utils.sparkevents import SparkEvents
+from sparkmagic.livyclientlib.command import Command
+from sparkmagic.livyclientlib.exceptions import DataFrameParseException, BadUserDataException
+
+import ast
+
+class SparkStoreCommand(Command):
+    def __init__(self, code, output_var, samplemethod=None, maxrows=None, samplefraction=None, spark_events=None):
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: It shouldn't take `code`

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+            command = u'collect({})'.format(command)
+        command = u'jsonlite::toJSON({})'.format(command)
+        command = u'for ({} in ({})) {{cat({})}}'.format(constants.LONG_RANDOM_VARIABLE_NAME,
+                                                         command,
+                                                         constants.LONG_RANDOM_VARIABLE_NAME)
+        return Command(command)
+
+
+    @staticmethod
+    def _records_to_dataframe(records_text, kind):
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: It should be easy to pull this method out into `utils.py` without much more refactoring and use it for this class and `SQLQuery`.

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        self.samplefraction = samplefraction
+        self.output_var = output_var
+        if spark_events is None:
+            spark_events = SparkEvents()
+        self._spark_events = spark_events
+
+    def execute(self, session):
+        return self.store_to_context(session)
+
+    def store_to_context(self, session):
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: This should just be in execute. No need to have this indirection.

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+            coerce_pandas_df_to_numeric_datetime(df)
+            return df
+        except ValueError:
+            raise DataFrameParseException(u"Cannot parse object as JSON: '{}'".format(strings))
+
+    # Used only for unit testing
+    def __eq__(self, other):
+        return self.code == other.code and \
+            hasattr(other, 'samplemethod') and \
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: Why the `hasattr`?

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+@with_setup(_setup, _teardown)
+def test_execute_code():
+    spark_events = MagicMock()
+    variable_name = "abc"
+    code = "hugedataset.collect()"
+    sparkcommand = SparkStoreCommand(code, variable_name, "take", 100, 0.2, spark_events=spark_events)
+    sparkcommand.to_command = MagicMock(return_value=MagicMock())
+    result = """{"z":100, "nullv":null, "y":50}
+{"z":25, "nullv":null, "y":10}"""
+    sparkcommand.to_command.return_value.execute = MagicMock(return_value=(True, result))
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: The 3 lines above are hard to parse 😛 
Can we introduce a `command` MagicMock variable. That way you can define what its `execute` method does more clearly and add it as the `return_value` of `to_command`.

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 22]
```diff
@@ -0,0 +1,241 @@
...
+    result_data = pd.DataFrame([{'z': 100, "nullv": None, 'y': 50}, {'z':25, "nullv":None, 'y':10}], columns=['z', "nullv", 'y'])
+    session = MagicMock()
+    session.kind = "pyspark"
+    result = sparkcommand.execute(session)
+    assert_frame_equal(result, result_data)
+    sparkcommand.to_command.assert_called_once_with(session.kind, variable_name)
+    sparkcommand.to_command.return_value.execute.assert_called_once_with(session)
+
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Nit: extra line

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 23]
```diff
@@ -0,0 +1,241 @@
...
+                  Command(u'for {} in {}.toJSON().take(120): print({}.encode("{}"))'\
+                          .format(LONG_RANDOM_VARIABLE_NAME, variable_name,
+                                  LONG_RANDOM_VARIABLE_NAME, conf.pyspark_python_encoding())))
+    assert_equals(sparkcommand._scala_command(variable_name),
+                  Command(u'{}.toJSON.take(120).foreach(println)'.format(variable_name)))
+
+
+@with_setup(_setup, _teardown)
+def test_pyspark_livy_sampling_options_spark2():
+        variable_name = "abc"
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: How are you specifying that these `SparkStoreCommand`s are Spark 2?

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+
+
+@with_setup(_setup, _teardown)
+def test_execute_code():
+    spark_events = MagicMock()
+    variable_name = "abc"
+    code = "hugedataset.collect()"
+    sparkcommand = SparkStoreCommand(code, variable_name, "take", 100, 0.2, spark_events=spark_events)
+    sparkcommand.to_command = MagicMock(return_value=MagicMock())
+    result = """{"z":100, "nullv":null, "y":50}
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Do we have unit tests for the code that transforms the strings to dataframes? I think we do... if so, we should move those tests to the utils tests when you move the function over. That way, you don't have to test that here, and you need only mock the utils function.

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 25]
```diff
@@ -0,0 +1,241 @@
...
+        sparkcommand = SparkStoreCommand(code, variable_name, samplemethod='sample', samplefraction=0.33, maxrows=3234)
+        assert_equals(sparkcommand._pyspark_command(variable_name),
+                      Command(u'for {} in {}.toJSON().sample(False, 0.33).take(3234): '
+                              u'print({}.encode("{}"))'\
+                              .format(LONG_RANDOM_VARIABLE_NAME, variable_name,
+                                      LONG_RANDOM_VARIABLE_NAME, conf.pyspark_python_encoding())))
+
+@with_setup(_setup, _teardown)
+def test_scala_livy_options_spark2():
+        variable_name = "abc"
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Ditto on Spark 2. Am I going blind? 👅

File: sparkmagic/sparkmagic/utils/configuration.py [hunk 26]
```diff
@@ -175,7 +175,10 @@ def default_samplefraction():
 def pyspark_sql_encoding():
     return u'utf-8'
     
-    
+@_with_override
+def pyspark_python_encoding():
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/utils/configuration.py]: Why not use the same sql encoding variable as above?

File: sparkmagic/sparkmagic/tests/test_kernel_magics.py (hunk 2 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_kernel_magics.py]: fixed

File: sparkmagic/sparkmagic/tests/test_kernel_magics.py (hunk 3 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_kernel_magics.py]: I assumed there is a reason it happens for sql right below. 
I agree, from the code there is no clear reason why this is happening. Removed the check

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 10 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: comment not clear. We are checking the vanilla-case, simple good old spark command; no storing, hence no storing class created.

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 9 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: same as above.

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 8 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: same for test_run_sql_command_exception() ? Similar tests exist in test_sparkmagicbase.py

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 7 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: same as above

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 4 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: Roger

File: sparkmagic/sparkmagic/tests/test_sparkmagicbase.py (hunk 5 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkmagicbase.py]: Agree; fixed.

File: sparkmagic/sparkmagic/tests/test_sparkmagicbase.py (hunk 11 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkmagicbase.py]: Added check that no SparkStoreCommand is created altogether

File: sparkmagic/sparkmagic/tests/test_sparkmagicbase.py (hunk 6 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkmagicbase.py]: Added

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 12 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: indeed, added also for invalid

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 13 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Fixed. Should this also be done in test_sqlquery.py ?

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 15 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Added

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 16 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: fixed (removed)

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 24 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Since we have additional tests for that, I tried to simplify the test by removing the additional test on the DF format.

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py (hunk 17 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: Fixed

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py (hunk 19 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: fixed

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py (hunk 18 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: Done

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py (hunk 20 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: SparkStoreCommand inherits from Command. assert_any_call and similar functions throw an exception when comparing instances of the two. Haven't checked if this is bug in nose or expected behavior.

File: sparkmagic/sparkmagic/utils/configuration.py (hunk 26 above)
Comment (orestis-ms) [inline sparkmagic/sparkmagic/utils/configuration.py]: clarity, independence?

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py (hunk 20 above)
Comment (aggFTW) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: Hmmm... didn't understand this. Want to chat on Monday?

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 10 above)
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: I'm saying that we don't need to check that the run_cell_method is called and only that the sparkmagicsbase `execute_spark` method is being called with the right parameters. This unit test is here only to check that parsing is correct.

File: sparkmagic/sparkmagic/tests/test_remotesparkmagics.py (hunk 8 above)
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_remotesparkmagics.py]: I think that there should be unit tests that test only what this class does. Since this class is not actually executing anything against the spark_controller when running spark (sparkmagicsbase is), this should not test an exception there. Instead, it should just test that the exception is caught when it is thrown by `execute_spark`. If this is not clear, we can discuss on Monday.

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 13 above)
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Yes

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+                  Command(u'for {} in {}.toJSON().take(120): print({}.encode("{}"))'\
+                          .format(LONG_RANDOM_VARIABLE_NAME, variable_name,
+                                  LONG_RANDOM_VARIABLE_NAME, conf.pyspark_python_encoding())))
+    assert_equals(sparkcommand._scala_command(variable_name),
+                  Command(u'{}.toJSON.take(120).foreach(println)'.format(variable_name)))
+
+
+@with_setup(_setup, _teardown)
+def test_scala_livy_options_spark2():
+    variable_name = "var_name"
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Where is Spark 2 here?

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 28]
```diff
@@ -0,0 +1,246 @@
...
+
+    sparkcommand = SparkStoreCommand(variable_name, samplemethod='sample', samplefraction=0.33, maxrows=3234)
+    assert_equals(sparkcommand._scala_command(variable_name),
+                  Command('{}.toJSON.sample(false, 0.33).take(3234).foreach(println)'.format(variable_name)))
+
+def test_sample_method_none_all_langs():
+    variable_name = "var_name"
+
+    sparkcommand = SparkStoreCommand(variable_name, samplemethod=None, maxrows=100)
+
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Too many blank lines.

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py [hunk 29]
```diff
@@ -0,0 +1,246 @@
...
+    sparkcommand = SparkStoreCommand(variable_name, samplemethod='sample', samplefraction=0.25, maxrows=-1)
+    assert_equals(sparkcommand._scala_command(variable_name),
+                  Command('{}.toJSON.sample(false, 0.25).collect.foreach(println)'.format(variable_name)))
+
+    sparkcommand = SparkStoreCommand(variable_name, samplemethod='sample', samplefraction=0.33, maxrows=3234)
+    assert_equals(sparkcommand._scala_command(variable_name),
+                  Command('{}.toJSON.sample(false, 0.33).take(3234).foreach(println)'.format(variable_name)))
+
+def test_sample_method_none_all_langs():
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: Can this just be added to the different language tests directly?

File: sparkmagic/sparkmagic/utils/configuration.py (hunk 26 above)
Comment (aggFTW) [inline sparkmagic/sparkmagic/utils/configuration.py]: Hmm... I don't think it's needed. If you want, you can call it `dataframe_python_encoding`. Clarity is OK, but I don't see how independence is ever needed for this, do you?

File: sparkmagic/sparkmagic/utils/utils.py [hunk 30]
```diff
@@ -53,6 +56,29 @@ def coerce_pandas_df_to_numeric_datetime(df):
             except (ValueError, TypeError):
                 pass
 
+
+def records_to_dataframe(records_text, kind):
```
Comment (aggFTW) [inline sparkmagic/sparkmagic/utils/utils.py]: Do we have tests for this? Should we move them to the utils test file?

Comment (orestis-ms) [inline sparkmagic/sparkmagic/utils/utils.py]: I'm noticing that there no tests for records_to_dataframe, but only for the sub-routine coerce_pandas_df_to_numeric_datetime. The latter is tested extensively, but not the former.

Comment (aggFTW) [inline sparkmagic/sparkmagic/utils/utils.py]: 😞 It seems like we were indirectly testing that function by having empty content, invalid content, and etcetera go through the function.

It's OK for now to leave it untested. Please just create an issue specifying that the function has no unit tests.

File: sparkmagic/sparkmagic/tests/test_sparkstorecommand.py (hunk 27 above)
Comment (aggFTW) [inline sparkmagic/sparkmagic/tests/test_sparkstorecommand.py]: You don't need Spark 2 tests because you are not referencing the name of the sql context, which changes from spark 1 to spark 2.

**Comment by aggFTW** [comment]
So, TODO:

- [ ] Did you test it on an actual notebook? Can you list the tests you performed?
- [ ] Create an issue on missing JSON to dataframe unit tests
- [ ] Bump library version

Do we have anything else?

File: autovizwidget/autovizwidget/__init__.py [hunk 31]
```diff
@@ -1 +1 @@
-__version__ = '0.10.4'
+__version__ = '0.10.5'
```
Comment (aggFTW) [inline autovizwidget/autovizwidget/__init__.py]: Please bump to 0.11.0

You can read about semantic versioning here: http://semver.org/
Since we haven't reached 1.0 yet, we are bumping minor version whenever we introduce a new feature, and bumping build version whenever we make a bug fix.

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py [hunk 32]
```diff
@@ -0,0 +1,120 @@
+from sparkmagic.utils.utils import coerce_pandas_df_to_numeric_datetime, records_to_dataframe
```
Comment (msftristew) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: Nit: Rename file to `sparkstorecommand.py`?

File: sparkmagic/sparkmagic/livyclientlib/sparkcommand.py [hunk 33]
```diff
@@ -0,0 +1,120 @@
...
+import sparkmagic.utils.configuration as conf
+import sparkmagic.utils.constants as constants
+from sparkmagic.utils.sparkevents import SparkEvents
+from sparkmagic.livyclientlib.command import Command
+from sparkmagic.livyclientlib.exceptions import DataFrameParseException, BadUserDataException
+
+import ast
+
+class SparkStoreCommand(Command):
```
Comment (msftristew) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: Structurally this class looks quite a lot like the `SqlQuery` class.  Can/should this be refactored?

Comment (orestis-ms) [inline sparkmagic/sparkmagic/livyclientlib/sparkcommand.py]: We have discussed offline about this. There are a couple of ways to go. The straight-forward is to refactor so that the new class and SQLQuery inherit from some hybrid. 

However. if we include Command, they are all 'commands', and all (directly or indirectly) inherit from ObjectWithGuid. Hence, one can argue that SQLQuery should inherit from Command.

**Comment by Nithanaroy** [comment]
This is amazing! Is there an example on how to use this to send a Scala Spark variable / dataframe to local Python in Spark kernel?

**Comment by philmassie** [comment]
@Nithanaroy this is a bit late, but I found this works:
```
%%spark -o pdf -n -1
pdf = your_spark_dataframe

pdf.head()
```
(watch out for -n -1)
